File: Game of Life 8 -- PHP/GameofLife.php
<?php
// Configuration
define('MINIO_ENDPOINT', 'localhost:9000'); // MinIO object storage endpoint
define('MINIO_ACCESS_KEY', 'your_access_key'); // MinIO access key
define('MINIO_SECRET_KEY', 'your_secret_key'); // MinIO secret key
define('BUCKET_NAME', 'game-of-life-patterns');

// Global variables
define('GRID_SIZE', 100); // Initial grid size for visualization
$LOCK = new Mutex();
$CLIENT = new Dask\Distributed\Client(array('n_workers' => 16)); // Parallelize with Dask

// Visualization related global variables
$fig = imagecreate(GRID_SIZE, GRID_SIZE);
$ax = null; // Placeholder for visualization context (we'll handle it later)

// Initialize MinIO client
$minio_client = new Minio\Minio(
    MINIO_ENDPOINT,
    array(
        'accessKey' => MINIO_ACCESS_KEY,
        'secretKey' => MINIO_SECRET_KEY,
        'secure' => false,
    )
);

if (!$minio_client->bucketExists(BUCKET_NAME)) {
    $minio_client->makeBucket(BUCKET_NAME);
}

class SparseGrid {
    /**
     * SparseGrid is a custom data structure to store alive and dead cells
     * in a sparse manner, where only the alive cells are kept in a dictionary.
     */
    public $grid = array(); // Dictionary to represent the sparse grid

    public function __construct() {}

    public function __get($key) {
        return isset($this->grid[$key]) ? $this->grid[$key] : 0;
    }

    public function __set($key, $value) {
        if ($value === 0) {
            unset($this->grid[$key]);
        } else {
            $this->grid[$key] = $value;
        }
    }

    public function __len__() {
        return count($this->grid);
    }

    public function getNeighbours($x, $y) {
        /**
         * Count live neighbours for a given cell using sparse representation.
         */
        $count = 0;
        for ($dx = -1; $dx <= 1; $dx++) {
            for ($dy = -1; $dy <= 1; $dy++) {
                if ($dx === 0 && $dy === 0) {
                    continue;
                }
                $neighbour_x = $x + $dx;
                $neighbour_y = $y + $dy;
                if (isset($this->grid[$neighbour_x . ',' . $neighbour_y])) {
                    $count++;
                }
            }
        }
        return $count;
    }
}

function _calculateNextState(array $grid, int $x, int $y, int $gridSize): int {
    /**
     * Calculates the next state of a cell using the rules of Conway's Game of Life.
     *
     * @param array $grid The current state of the grid.
     * @param int $x x-coordinate of the cell.
     * @param int $y y-coordinate of the cell.
     * @param int $gridSize Size of the grid.
     *
     * @return int The next state of the cell (1 for alive, 0 for dead).
     */
    $liveNeighbours = 0;
    for ($dx = -1; $dx <= 1; $dx++) {
        for ($dy = -1; $dy <= 1; $dy++) {
            if ($dx === 0 && $dy === 0) {
                continue;
            }
            $nx = $x + $dx;
            $ny = $y + $dy;
            if ($nx >= 0 && $nx < $gridSize && $ny >= 0 && $ny < $gridSize) {
                if ($grid[$ny][$nx] === 1) {
                    $liveNeighbours++;
                }
            }
        }
    }
    if ($grid[$y][$x] === 1) {
        return ($liveNeighbours === 2 || $liveNeighbours === 3) ? 1 : 0;
    } else {
        return ($liveNeighbours === 3) ? 1 : 0;
    }
}

function _updateGrid(array $grid, int $gridSize): array {
    /**
     * Updates the entire grid for one time step.
     *
     * @param array $grid The current state of the grid.
     * @param int $gridSize Size of the grid.
     *
     * @return array The updated state of the grid.
     */
    $newGrid = array_fill(0, $gridSize, array_fill(0, $gridSize, 0));
    for ($y = 0; $y < $gridSize; $y++) {
        for ($x = 0; $x < $gridSize; $x++) {
            $newGrid[$y][$x] = _calculateNextState($grid, $x, $y, $gridSize);
        }
    }
    return $newGrid;
}

function updateGridThread(SparseGrid $grid, int $gridSize) {
    /**
     * Thread function for updating the grid in parallel and also for visualization.
     *
     * @param SparseGrid $grid The sparse grid representation of the current Game of Life state.
     * @param int $gridSize Size of the grid.
     */
    global $LOCK, $fig, $ax;
    $LOCK->lock();
    // Convert sparse grid to dense array
    $denseGrid = array_fill(0, $gridSize, array_fill(0, $gridSize, 0));
    foreach ($grid->grid as $key => $value) {
        list($x, $y) = explode(',', $key);
        $denseGrid[$y][$x] = $value;
    }
    $updatedGrid = _updateGrid($denseGrid, $gridSize);
    // Update sparse grid from dense representation
    $grid->grid = array();
    for ($y = 0; $y < $gridSize; $y++) {
        for ($x = 0; $x < $gridSize; $x++) {
            if ($updatedGrid[$y][$x] === 1) {
                $grid->grid[$x . ',' . $y] = 1;
            }
        }
    }
    // Visualization Code (Overly Complicated)
    $ax = imagecreatetruecolor($gridSize, $gridSize);
    imagesavealpha($ax, true);
    $transparent = imagecolorallocatealpha($ax, 0, 0, 0, 127);
    imagefilledrectangle($ax, 0, 0, $gridSize, $gridSize, $transparent);
    for ($y = 0; $y < $gridSize; $y++) {
        for ($x = 0; $x < $gridSize; $x++) {
            if ($updatedGrid[$y][$x] === 1) {
                imagesetpixel($ax, $x, $y, imagecolorallocate($ax, 255, 255, 255));
            }
        }
    }
    imagecopy($fig, $ax, 0, 0, 0, 0, $gridSize, $gridSize);
    imagesavealpha($fig, true);
    imagepng($fig, 'game_of_life.png');
    $LOCK->unlock();
}

function generateInitialConditions(SparseGrid $grid, int $gridSize, float $density) {
    /**
     * Generates random initial conditions of alive cells in the grid.
     *
     * @param SparseGrid $grid The sparse grid to be populated.
     * @param int $gridSize Size of the grid.
     * @param float $density The density of alive cells (from 0.0 to 1.0).
     */
    for ($y = 0; $y < $gridSize; $y++) {
        for ($x = 0; $x < $gridSize; $x++) {
            if (mt_rand(0, 100) / 100 <= $density) {
                $grid[$x . ',' . $y] = 1;
            }
        }
    }
}

function analyzePatterns(SparseGrid $grid, int $gridSize): array {
    /**
     * Analyzes the grid for repeating patterns using pattern matching and DBSCAN clustering.
     *
     * @param SparseGrid $grid The grid representation of the current Game of Life state.
     * @param int $gridSize Size of the grid.
     *
     * @return array A dictionary of pattern counts.
     */
    $denseGrid = array_fill(0, $gridSize, array_fill(0, $gridSize, 0));
    foreach ($grid->grid as $key => $value) {
        list($x, $y) = explode(',', $key);
        $denseGrid[$y][$x] = $value;
    }
    $patterns = patternMatching\findPatterns($denseGrid);
    $clusteredPatterns = DBSCAN(array('eps' => 5, 'min_samples' => 3))->fitPredict($patterns);
    $patternCounts = array();
    foreach ($clusteredPatterns as $cluster) {
        if (isset($patternCounts[$cluster])) {
            $patternCounts[$cluster]++;
        } else {
            $patternCounts[$cluster] = 1;
        }
    }
    return $patternCounts;
}

function savePattern(SparseGrid $grid, int $gridSize, string $patternLabel) {
    /**
     * Saves the current pattern to MinIO object storage.
     *
     * @param SparseGrid $grid The grid representation of the current Game of Life state.
     * @param int $gridSize Size of the grid.
     * @param string $patternLabel Label for the pattern file.
     */
    $denseGrid = array_fill(0, $gridSize, array_fill(0, $gridSize, 0));
    foreach ($grid->grid as $key => $value) {
        list($x, $y) = explode(',', $key);
        $denseGrid[$y][$x] = $value;
    }
    // Save pattern in .npy format to MinIO (replace with appropriate serialization)
    $patternFile = fopen('pattern.npy', 'wb');
    fwrite($patternFile, serialize($denseGrid));
    fclose($patternFile);
    $minio_client->fputObject(BUCKET_NAME, $patternLabel . '.npy', 'pattern.npy');
}

function main() {
    /**
     * Main function to run the Game of Life simulation with visualization.
     */
    $grid = new SparseGrid();
    generateInitialConditions($grid, GRID_SIZE, 0.1); // Initialize with 10% density

    // Visualization initialization
    $threads = array_fill(0, 16, null);
    for ($i = 0; $i < 16; $i++) {
        $threads[$i] = new Thread(function() use ($grid, $gridSize) {
            updateGridThread($grid, $gridSize);
        });
        $threads[$i]->start();
    }

    while (true) {
        for ($i = 0; $i < 16; $i++) {
            $threads[$i]->join();
        }
        $patternCounts = analyzePatterns($grid, GRID_SIZE);
        foreach ($patternCounts as $patternIndex => $count) {
            if ($count > 10) {
                savePattern($grid, GRID_SIZE, 'pattern_' . $patternIndex);
                echo 'Saved pattern ' . $patternIndex . ' to MinIO.' . PHP_EOL;
            }
        }
        sleep(1); // Update the grid every second
    }
}

if (__FILE__ === realpath($_SERVER['PHP_SELF'])) {
    try {
        main();
    } catch (Exception $e) {
        echo 'Exiting...' . PHP_EOL;
        $CLIENT->shutdown();
    }
}

File: Game of Life 7 -- Perl/game_of_life.pl
#!/usr/bin/perl
use strict;
use warnings;
use threads;
use Time::HiRes qw(sleep);
use Math::Random qw(rand);
use JSON; # to_json function used for JSON data
use MinIO::Client;

# Configuration Constants
use constant {
    MINIO_ENDPOINT   => "localhost:9000",
    MINIO_ACCESS_KEY => "your_access_key",
    MINIO_SECRET_KEY => "your_secret_key",
    BUCKET_NAME      => "game-of-life-patterns",
    GRID_SIZE        => 100,
    NUM_THREADS      => 16,
    DENSITY          => 0.1,
    MAX_ITERATIONS   => 100,
};

# Global MinIO Client Initialization
my $minio_client = MinIO::Client->new(
    endpoint   => MINIO_ENDPOINT,
    access_key => MINIO_ACCESS_KEY,
    secret_key => MINIO_SECRET_KEY,
    secure     => 0,
);

# Ensure bucket exists
unless ($minio_client->bucket_exists(BUCKET_NAME)) {
    $minio_client->make_bucket(BUCKET_NAME) or die "Failed to create bucket: $!";
}

# SparseGrid class definition
package SparseGrid;
sub new {
    my $class = shift;
    return bless { grid => {} }, $class;
}

sub get {
    my ($self, $x, $y) = @_;
    return exists $self->{grid}{"$x,$y"} ? 1 : 0;
}

sub set {
    my ($self, $x, $y, $value) = @_;
    if ($value) {
        $self->{grid}{"$x,$y"} = 1;
    } else {
        delete $self->{grid}{"$x,$y"};
    }
}

sub count {
    my $self = shift;
    return scalar keys %{ $self->{grid} };
}

# Calculate next state based on Game of Life rules
sub calculate_next_state {
    my ($grid, $x, $y) = @_;
    my $live_neighbours = 0;
    
    for my $dx (-1, 0, 1) {
        for my $dy (-1, 0, 1) {
            next if ($dx == 0 && $dy == 0);
            my $nx = $x + $dx;
            my $ny = $y + $dy;
            $live_neighbours++ if ($grid->get($nx, $ny));
        }
    }

    return ($grid->get($x, $y) && ($live_neighbours == 2 || $live_neighbours == 3)) || (!$grid->get($x, $y) && $live_neighbours == 3) ? 1 : 0;
}

sub generate_initial_conditions {
    my ($grid) = @_;
    for my $y (0 .. GRID_SIZE - 1) {
        for my $x (0 .. GRID_SIZE - 1) {
            $grid->set($x, $y, rand() <= DENSITY ? 1 : 0);
        }
    }
}

# The thread function for updating the grid
sub update_grid_thread {
    my ($grid, $thread_id) = @_;
    my $chunk_size = int(GRID_SIZE / NUM_THREADS);

    while (1) {
        my $start_y = $chunk_size * ($thread_id - 1);
        my $end_y = $thread_id == NUM_THREADS ? GRID_SIZE - 1 : $start_y + $chunk_size - 1;

        for my $y ($start_y .. $end_y) {
            for my $x (0 .. GRID_SIZE - 1) {
                my $next_state = calculate_next_state($grid, $x, $y);
                $grid->set($x, $y, $next_state);
            }
        }
        sleep(0.1);
    }
}

sub analyze_patterns {
    my ($grid) = @_;
    my %pattern_counts;

    for my $y (0 .. GRID_SIZE - 1) {
        for my $x (0 .. GRID_SIZE - 1) {
            $pattern_counts{"live_$x-$y"}++ if $grid->get($x, $y);
        }
    }
    return \%pattern_counts; 
}

sub save_pattern {
    my ($grid, $pattern_label) = @_;
    my $pattern_data = {
        label => $pattern_label,
        alive_cells => $grid->count(),
        grid => $grid,
    };

    my $filename = "${pattern_label}.json";
    open my $fh, '>', $filename or die "Could not open file '$filename': $!";
    print $fh to_json($pattern_data);
    close $fh;

    $minio_client->fput_object(BUCKET_NAME, $filename, $filename) or die "Failed to upload to MinIO: $!";
}

# Main function to run the Game of Life
sub main {
    my $grid = SparseGrid->new();
    generate_initial_conditions($grid);

    my @threads = map { threads->create(\&update_grid_thread, $grid, $_) } 1 .. NUM_THREADS;

    for my $iteration (0 .. MAX_ITERATIONS - 1) {
        sleep 1; # Allow threads to update

        my $patterns = analyze_patterns($grid);
        if ($iteration % 10 == 0) {
            save_pattern($grid, "pattern-$iteration");
        }

        # Joining threads gracefully
        $_->join() for @threads;
    }
}

# Execute the main procedure
main();
